### ISSUE TITLE
All read calls [ORecordId.getRecord()] takes long time when commit() is happening in another Thread

### ISSUE THREAD

**Issue by adithyank** [orientechnologies/orientdb#3283]
OrientDB Version : 2.0-M3

In our application, we do all writing in one Thread and read in multiple threads.

Pls find below the thread dump of the writing thread....
```
"InventoryDataPersister" prio=10 tid=0x00007ef473ff8800 nid=0x1f42 runnable [0x00007ef4fafeb000]
   java.lang.Thread.State: RUNNABLE
        at sun.misc.Unsafe.arrayBaseOffset(Native Method)
        at com.orientechnologies.common.directmemory.OUnsafeMemoryJava7.get(OUnsafeMemoryJava7.java:31)
        at com.orientechnologies.common.directmemory.ODirectMemoryPointer.get(ODirectMemoryPointer.java:79)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.base.ODurablePage.setIntValue(ODurablePage.java:126)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.OClusterPage.appendRecord(OClusterPage.java:136)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.OPaginatedCluster.addEntry(OPaginatedCluster.java:1375)
        at com.orientechnologies.orient.core.storage.impl.local.paginated.OPaginatedCluster.createRecord(OPaginatedCluster.java:387)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.doCreateRecord(OAbstractPaginatedStorage.java:1359)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.commitEntry(OAbstractPaginatedStorage.java:1832)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.commit(OAbstractPaginatedStorage.java:860)
        at com.orientechnologies.orient.core.tx.OTransactionOptimistic.doCommit(OTransactionOptimistic.java:455)
        at com.orientechnologies.orient.core.tx.OTransactionOptimistic.commit(OTransactionOptimistic.java:142)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(ODatabaseDocumentTx.java:2261)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.commit(ODatabaseDocumentTx.java:2231)
        at in.co.nmsworks.documenttdb.db.DocDB.commit(DocDB.java:482)
        at in.co.nmsworks.cygnet.telecom.fault.rcasia.inventory.store.RcaInvWriter.addAll(RcaInvWriter.java:118)
        at in.co.nmsworks.cygnet.telecom.fault.rcasia.inventory.store.RcaInvWriter.addAll(RcaInvWriter.java:76)
        at in.co.nmsworks.cygnet.telecom.fault.rcasia.inventory.dao.InventoryDAO.save(InventoryDAO.java:106)
        at in.co.nmsworks.cygnet.telecom.fault.rcasia.inventory.queue.InventoryDataPersister.saveDAOData(InventoryDataPersister.java:331)
        at in.co.nmsworks.cygnet.telecom.fault.rcasia.inventory.queue.InventoryDataPersister.processAddition(InventoryDataPersister.java:306)
        at in.co.nmsworks.cygnet.telecom.fault.rcasia.inventory.queue.InventoryDataPersister.run(InventoryDataPersister.java:71)

   Locked ownable synchronizers:
        - <0x00007efaa7385a30> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7385d30> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7386000> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7387950> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7387ce0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7387d70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7387dd0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7387ef0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7387fb0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388070> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388550> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa73885b0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388640> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa73886a0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388910> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388a60> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388b20> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388b50> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388be0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388c70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388ca0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388eb0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7388ee0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535180> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75351b0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535330> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535450> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535570> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535690> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535950> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535b00> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535d70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535dd0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535ef0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7535fe0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536130> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75361f0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536340> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75363a0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536430> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536490> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536610> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536700> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75368b0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536ac0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536be0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536c70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536d90> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536f40> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7536f70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7538ad0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7538d40> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7538e00> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7538e30> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7538f20> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7538f50> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7538fe0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7539040> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7539130> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75391f0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7539220> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7539310> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75393a0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75394f0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7539760> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75398e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75399a0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7539bb0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7539cd0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7539ee0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a060> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a150> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a2a0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a2d0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a570> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a720> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a7e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a8d0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a990> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753a9f0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753adb0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753ae70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753b080> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753b110> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753b290> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753b380> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753b410> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753b560> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753b6e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa753b7d0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7545570> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7545660> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75456f0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7545720> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7545960> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7545b70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7545db0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7545e40> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7545e70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75460e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7546200> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7546380> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75463e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7546410> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7546590> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7546650> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75466e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7546a70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754cd10> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754cd40> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754cd70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d010> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d190> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d1c0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d310> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d3a0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d3d0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d550> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d5b0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d8b0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754d8e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754da90> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754dca0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754dd30> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754dd90> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754ddc0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754e000> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754e090> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754f210> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754f2a0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754f2d0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754f360> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754f540> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754f780> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754f8a0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754f990> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754f9c0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754fb40> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754fc00> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754fc60> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754fcf0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754fe40> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754fe70> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754ff00> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa754ff90> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75501d0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7550200> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75502c0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7550350> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7550380> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75503e0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7550470> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa7550710> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa756b530> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa756b9b0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa756bb30> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa756bbf0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa756bc20> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa756bcb0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa756bf20> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa756c0d0> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa756c100> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efaa75ad480> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        - <0x00007efab6e61ba8> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
````        

These are around 50 reading threads and all threads' thread dump is like below...

```
        "PlugInvokationTask-1041425-92935-CircuitPlug" daemon prio=10 tid=0x00007ef59eea9800 nid=0x29cc waiting on condition [0x00007ef2cdfba000]
   java.lang.Thread.State: TIMED_WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x00007efaa75ad480> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
        at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:226)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedNanos(AbstractQueuedSynchronizer.java:1033)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1326)
        at java.util.concurrent.locks.ReentrantReadWriteLock$ReadLock.tryLock(ReentrantReadWriteLock.java:873)
        at com.orientechnologies.common.concur.resource.OSharedResourceAdaptive.acquireSharedLock(OSharedResourceAdaptive.java:145)
        at com.orientechnologies.common.concur.resource.OSharedResourceAdaptiveExternal.acquireSharedLock(OSharedResourceAdaptiveExternal.java:44)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.readRecord(OAbstractPaginatedStorage.java:1198)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.readRecord(OAbstractPaginatedStorage.java:644)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeReadRecord(ODatabaseDocumentTx.java:1497)
        at com.orientechnologies.orient.core.tx.OTransactionOptimistic.loadRecord(OTransactionOptimistic.java:221)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:1354)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:123)
        at com.orientechnologies.orient.core.id.ORecordId.getRecord(ORecordId.java:309)
        at in.co.nmsworks.documenttdb.db.DocDB.getDocument(DocDB.java:204)
        at in.co.nmsworks.documenttdb.db.DocDB.getDocTypeObject(DocDB.java:152)
        at in.co.nmsworks.documenttdb.db.DocDB.getDocTypeObject(DocDB.java:147)
        at in.co.nmsworks.cygnet.telecom.fault.rcasia.inventory.store.RcaInvReader.getEntity(RcaInvReader.java:761)
        at in.co.nmsworks.cygnet.telecom.fault.rcasia.inventory.store.RcaInvReader.getEntity(RcaInvReader.java:751)
        at in.co.nmsworks.cygnet.telecom.fault.rcasia.inventory.store.RcaInvReader.getEntities(RcaInvReader.java:775)
```

Both writing time and reading time increases if both read and write is happening in multiple threads...?? If only the writing is happening, writing happens in 4 seconds for one transaction of ours. If reading is also happening in the same time, then, writing time is more than 10 seconds for the same volume of transaction. and all the read threads also go to TIMED_WAITING mode.

1. Is it the behavior as per design?
2. In this case, accessing OrientDB's same database in multiple threads is not possible with the same performance of single threaded one?
3. Or what else should be done?

**Comment by sajalgupta1976** [comment]
Hi, I am currently evaluating OrientDB for a use case involving concurrent read/write. It looks like locks are acquired at storage level(unless I am wrong) which can introduce bottleneck in a highly concurrent system involving writes. Any update on this issue would really help.
 BTW really like OrientDB for its great features. Thanks

**Comment by lvca** [comment]
Hey guys, we have a storage level locking on transaction commit. This means that transactions don't go in parallel, while single atomic operation yes. This is something we're working for and we should be able to release parallel TXs soon, maybe in 2.1 planned for February 2015.

**Comment by sajalgupta1976** [comment]
@lvca Thanks, In the thread dumps above, all the reads are also getting blocked during commit due to read lock which is acquired during read,Is this also something which is going to get fixed in 2.1 ? 
In the meantime is there any workaround/option to avoid such blocks during commits? 
Also I see the lock is opened with nofair option, wouldn't it be better to create the lock with fair option?

**Comment by lvca** [comment]
@laa Why locks around are not fair?

**Comment by adithyank** [comment]
Does the date of Febrary 2015 for release 2.1 still hold good?

**Comment by lvca** [comment]
2.1 will be release on February for sure. We don't know if Parallel TX could be done for 2.1 or postponed to the next release because it needs a complex work in the engine.

**Comment by laa** [comment]
@lvca about fair sync usally it includes a lot of thread paring, so latency will be noticable higher. But I think we can include it as special storage attribute for  the cases like this one.
WDYT ?

**Comment by lvca** [comment]
Super idea! Like:

    ALTER DATABASE LOCK FAIR

?

**Comment by laa** [comment]
:) yes, something like this.

On Fri, Jan 23, 2015 at 4:55 PM, Luca Garulli <notifications@github.com>
wrote:

> Super idea! Like:
>
> ALTER DATABASE LOCK FAIR
>
> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/orientechnologies/orientdb/issues/3283#issuecomment-71204299>
> .
>



-- 
Best regards,
Andrey Lomakin.

**Comment by lvca** [comment]
Ok, I'm opening a new issue...

**Comment by lvca** [comment]
https://github.com/orientechnologies/orientdb/issues/3450

**Comment by sajalgupta1976** [comment]
Hi,
I believe this is only going to fix the problem to certain extent.Most production deployments are multithreaded with high volumes of concurrent reads and if we are using graphs we cannot avoid concurrent commits during reads (issue described above).I think we do need to consider the following please:-
1) Reads and commits blocking one another:- Read should be lock free at storage and cluster level and should acquire only shared locks at record/page level. It should be able to construct a read consistent view and return results instead of waiting for the ongoing transactions/writes. (A very common NoSQL usecase of one writer and multiple readers)

2) Write/Commits blocking one another:- Again locks should be acquired at record level.Currently writes are locked at cluster level and commits are at storage level which is also blocking reads.

I do acknowledge that this is not a very straight forward thing to implement,but I think this is required to have OrientDB perform as good as single threaded setup.

**Comment by laa** [comment]
Yes,
Sure that is why we are going to implement:

   1. Parallel txs.
   2. Rid off global locks and use locks only on disk cache page level.


Also final implementation of parllel txs will block only on very short
interval and only pages are participated in current tx. Exclusive block
will be acquired only when you copy all your changes from wal log to disk
cache. during last phase of tx.

On Mon, Feb 2, 2015 at 9:45 AM, sajalgupta1976 <notifications@github.com>
wrote:

> Hi,
> I believe this is only going to fix the problem to certain extent.Most
> production deployments are multithreaded with high volumes of concurrent
> reads and if we are using graphs we cannot avoid concurrent commits during
> reads (issue described above).I think we do need to consider the following
> please:-
> 1) Reads and commits blocking one another:- Read should be lock free at
> storage and cluster level and should acquire only shared locks at
> record/page level. It should be able to construct a read consistent view
> and return results instead of waiting for the ongoing transactions/writes.
> (A very common NoSQL usecase of one writer and multiple readers)
>
> 2) Write/Commits blocking one another:- Again locks should be acquired at
> record level.Currently writes are locked at cluster level and commits are
> at storage level which is also blocking reads.
>
> I do acknowledge that this is not a very straight forward thing to
> implement,but I think this is required to have OrientDB perform as good as
> single threaded setup.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/orientechnologies/orientdb/issues/3283#issuecomment-72417033>
> .
>



-- 
Best regards,
Andrey Lomakin.

**Comment by kaaljabr** [comment]
I am on 2.0.2 both DB and API. And I am facing this issue:
`````
in context with path [/aa-sport] threw exception [com.orientechnologies.orient.core.exception.ODatabaseException: Error on retrieving record #27:165831 (cluster: event)] with root cause
com.orientechnologies.common.concur.OTimeoutException: Can not lock record for 2000 ms. seems record is deadlocked by other record
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.acquireReadLock(OAbstractPaginatedStorage.java:1308)
        at com.orientechnologies.orient.core.tx.OTransactionAbstract.lockRecord(OTransactionAbstract.java:120)
        at com.orientechnologies.orient.core.id.ORecordId.lock(ORecordId.java:282)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.lockRecord(OAbstractPaginatedStorage.java:1784)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.readRecord(OAbstractPaginatedStorage.java:1424)
        at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.readRecord(OAbstractPaginatedStorage.java:697)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.executeReadRecord(ODatabaseDocumentTx.java:1572)
        at com.orientechnologies.orient.core.tx.OTransactionNoTx.loadRecord(OTransactionNoTx.java:80)
        at com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx.load(ODatabaseDocumentTx.java:1437)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.readRecord(ONetworkProtocolBinary.java:1453)
        at com.orientechnologies.orient.server.network.protocol.binary.ONetworkProtocolBinary.executeRequest(ONetworkProtocolBinary.java:346)
        at com.orientechnologies.orient.server.network.protocol.binary.OBinaryNetworkProtocolAbstract.execute(OBinaryNetworkProtocolAbstract.java:216)
        at com.orientechnologies.common.thread.OSoftThread.run(OSoftThread.java:65)
`````

When writing and reading at the same time. Is this a related issue? And how to fix it please. I didn't have this issue in 1.7.10

**Comment by adithyank** [comment]
Where are we in this issue???
Has it been resolved? are any possibility of resolving in near future?

**Comment by laa** [comment]
Hi it is duplication of https://github.com/orientechnologies/orientdb/issues/1677 . As you can see it is already in progress. Also it is partially solved in 2.2 version.
